presto-main/src/main/java/com/facebook/presto/tz/JdkBasedDateTimeZone.java

```diff
@@ -0,0 +1,114 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tz;
+
+import org.joda.time.DateTimeZone;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.zone.ZoneOffsetTransition;
+import java.time.zone.ZoneRules;
+import java.util.Objects;
+
+public class JdkBasedDateTimeZone
+        extends DateTimeZone
+{
+    private final ZoneRules zoneRules;
+
+    public JdkBasedDateTimeZone(String zoneId)
+    {
+        super(zoneId);
+        this.zoneRules = ZoneId.of(zoneId).getRules();
+    }
+
+    @Override
+    public String getNameKey(long instant)
+    {
+        // It is ok to return a dummy value here.
+        // getNameKey is used in 6 instances in Joda
+        // * `DateTimeZone.getShortName`. Assuming DefaultNameProvider is being used,
+        //   it then invokes `DefaultNameProvider.getShortName`, which in turn invokes
+        //   `DefaultNameProvider.getNameSet`, which ignores the passed in nameKey,
+        //   other than validating that it's not null.
+        // * `DateTimeZone.getName`. Same as above.
+        // * `CachedDateTimeZone.Info.getNameKey`. It is used to implement getNameKey.
+        // * `DateTimeZoneBuilder.PrecalculatedZone.getNameKey`. Same as above.
+        // * `DateTimeZoneBuilder.writeTo`. It only uses getNameKey when the object
+        //   is an instance of `FixedDateTimeZone`.
+        // * `ZoneInfoCompiler.test`. It's only invoked with built-in implementations
+        //   of `DateTimeZone`.
+
+        // Use an obviously incorrect and easily searchable value here
+        return "presto-name-key-not-provided";
+    }
+
+    @Override
+    public int getOffset(long instant)
+    {
+        return zoneRules.getOffset(Instant.ofEpochMilli(instant)).getTotalSeconds() * 1000;
+    }
+
+    @Override
+    public int getStandardOffset(long instant)
+    {
+        return zoneRules.getStandardOffset(Instant.ofEpochMilli(instant)).getTotalSeconds() * 1000;
+    }
+
+    @Override
+    public boolean isFixed()
+    {
+        return zoneRules.isFixedOffset();
+    }
+
+    @Override
+    public long nextTransition(long instant)
+    {
+        ZoneOffsetTransition nextTransition = zoneRules.nextTransition(Instant.ofEpochMilli(instant));
+        if (nextTransition == null) {
+            // this is after the last transition
+            return instant;
+        }
+        return nextTransition.toEpochSecond() * 1000;
+    }
+
+    @Override
+    public long previousTransition(long instant)
+    {
+        ZoneOffsetTransition previousTransition = zoneRules.previousTransition(Instant.ofEpochMilli(instant + 1));
+        if (previousTransition == null) {
+            // this is before the first transition
+            return instant;
+        }
+        return previousTransition.toEpochSecond() * 1000 - 1;
```

Comment:
In addition to addressing your comments, I added `+ 1` and `- 1` here. I also added a bunch of tests. Please take a look.

Reply:
Please add a comment here.